vm/systemSmartContracts/liquidStaking.go

```diff
@@ -199,25 +306,256 @@ func (l *liquidStaking) reDelegateRewardsFromPosition(args *vmcommon.ContractCal
-	returnCode := l.checkArgumentsWhenPositionIsInput(args)
+func (l *liquidStaking) burnAndExecuteFromESDTTransfer(
+	callerAddr []byte,
+	esdtTransfer *vmcommon.ESDTTransfer,
+	functionToCall string,
+) (*LiquidStakingAttributes, [][]byte, vmcommon.ReturnCode) {
+	attributes, err := l.getAttributesForNonce(esdtTransfer.ESDTTokenNonce)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return nil, nil, vmcommon.UserError
+	}
+
+	err = l.burnNFT(esdtTransfer.ESDTTokenNonce, esdtTransfer.ESDTValue)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return nil, nil, vmcommon.UserError
+	}
+
+	returnData, returnCode := l.executeOnDestinationSC(
+		attributes.ContractAddress,
+		functionToCall,
+		callerAddr,
+		esdtTransfer.ESDTValue,
+		attributes.RewardsCheckpoint,
+	)
 	if returnCode != vmcommon.Ok {
-		return returnCode
+		return nil, nil, returnCode
 	}
 
-	return vmcommon.Ok
+	return attributes, returnData, vmcommon.Ok
+}
+
+func (l *liquidStaking) executeOnDestinationSC(
+	dstSCAddress []byte,
+	functionToCall string,
+	userAddress []byte,
+	valueToSend *big.Int,
+	rewardsCheckPoint uint32,
+) ([][]byte, vmcommon.ReturnCode) {
+	txData := functionToCall + "@" + hex.EncodeToString(userAddress) + "@" + hex.EncodeToString(valueToSend.Bytes())
+	if rewardsCheckPoint > 0 {
+		txData += "@" + hex.EncodeToString(big.NewInt(int64(rewardsCheckPoint)).Bytes())
+	}
+	vmOutput, err := l.eei.ExecuteOnDestContext(dstSCAddress, l.liquidStakingSCAddress, big.NewInt(0), []byte(txData))
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return nil, vmcommon.UserError
+	}
+
+	if vmOutput.ReturnCode != vmcommon.Ok {
+		return nil, vmOutput.ReturnCode
+	}
+
+	return vmOutput.ReturnData, vmcommon.Ok
+}
+
+func (l *liquidStaking) createOrAddNFT(
+	delegationSCAddress []byte,
+	rewardsCheckpoint uint32,
+	value *big.Int,
+) (uint64, error) {
+	attributes := &LiquidStakingAttributes{
+		ContractAddress:   delegationSCAddress,
+		RewardsCheckpoint: rewardsCheckpoint,
+	}
+
+	marshalledData, err := l.marshalizer.Marshal(attributes)
+	if err != nil {
+		return 0, err
+	}
+
+	hash := l.hasher.Compute(string(marshalledData))
+	attrNonceKey := append([]byte(attributesNoncePrefix), hash...)
+	storageData := l.eei.GetStorage(attrNonceKey)
+	if len(storageData) > 0 {
+		nonce := big.NewInt(0).SetBytes(storageData).Uint64()
+		err = l.addQuantityToNFT(nonce, value)
+		if err != nil {
+			return 0, err
+		}
+
+		return nonce, nil
+	}
+
+	nonce, err := l.createNewNFT(value)
+	if err != nil {
+		return 0, nil
+	}
+
+	nonceBytes := big.NewInt(0).SetUint64(nonce).Bytes()
+	l.eei.SetStorage(attrNonceKey, nonceBytes)
+
+	nonceKey := append([]byte(nonceAttributesPrefix), nonceBytes...)
+	l.eei.SetStorage(nonceKey, marshalledData)
+
+	return nonce, nil
+}
+
+func (l *liquidStaking) createNewNFT(value *big.Int) (uint64, error) {
+	valuePlusOne := big.NewInt(0).Add(value, big.NewInt(1))
```

Comment:
why plus one?

Reply:
this a trick. You need to have at least 1 - in order to call AddQuantity - otherwise AddQuantity will fail. Thus we create one more and keep it in the contract.

Reply:
ok